### ISSUE TITLE
Trainer v2: cues + routines, spoken dynamic guidance, and a projection view you label categories in

### ISSUE THREAD

**Issue by thorwhalen** [thorwhalen/thoremin#163]
The v1 trainer (#160, merged in PR #162, live) proved the pipeline. This is the redesign the maintainer actually wants. **Supersedes the fixed four-step ritual**; keeps the maths.

---

## 1. Naming — the term to use

The maintainer asked for a better word than "training item". Proposal:

- **Cue** — one thing the system asks you to do ("look to your left", "hold any face and move closer"). A cue is already a performance instruction in music, it is short, and it works for both the spoken and the written form.
- **Routine** — a saved, ordered, named collection of cues you can run end to end.

Rejected: **prompt** (hopelessly overloaded with LLMs in this codebase — the assistant, the AI tool surface), **drill** (that is `tetrachord`'s term for a different thing and the two repos are read by the same people), **step** (what v1 called it; too anonymous to be a collection member), **exercise** (implies the player is practising, but here the *system* is the one learning).

## 2. What v1 has, and what to keep

| module | verdict |
|---|---|
| `src/enroll/sampler.ts` | **keep** — still-point gating is orthogonal to cues |
| `src/enroll/cluster.ts` | **keep** — hierarchy + `cutAt(k)`; becomes one of two ways to get categories |
| `src/enroll/invariance.ts` | **keep** — nuisance weighting is per-cue now, not per-ritual |
| `src/enroll/classify.ts` | **keep** — soft membership, reject, hysteresis |
| `src/enroll/ritual.ts` | **replace** — four hardcoded steps become a Cue collection |
| `src/enroll/session.ts` | **rework** into the Runner (below) |
| `src/app/TrainerPanel.tsx` | **rework** — becomes cue-picker + runner + projection view |
| `src/app/enroll/liveVector.ts` | **keep** — the DAG tap; already merges face + hand |

The whole thing is defined over `FeatureVector` (`Record<featureId, number>`), and that must not change: **246 catalog features today, 113 face and 133 hand.** A cue declares which feature *groups* it needs; the runner never learns which modality they came from. That is the extension seam for other signals and it already exists — do not add a face-specific type anywhere.

## 3. Cues as a zodal collection

A `Cue` is data (Zod schema), stored in a **`createNamedCollectionStore`** (`src/settings/namedCollection.ts`) — the same facade that already backs saved lab views (`src/app/lab/labViews.ts`) and named instruments. That gives CRUD, a `DataProvider` target (localStorage now, swappable later), and a summary list for free.

Roughly:

```ts
interface Cue {
  id: string; name: string;
  instruction: string;        // the written form, always shown
  voiceClip?: string;         // asset id of the cached TTS (see 4)
  collects: { groups: string[]; axes?: Invariance[] };
  sufficiency: SufficiencyRef; // how we know we have enough (see 5)
  variations?: string[];       // "again, a bit slower", "now the other way"
}
```

- **Add/remove is a dev+agent action for now, not a user one.** No user accounts. But the collection facade is the same one that would later serve users, so nothing has to be rebuilt.
- A **Routine** is a second collection whose payload is an ordered list of cue ids. Ship the Cue collection first; Routine is a thin second store over the same facade.
- The picker needs **filtering** — 246 features and a growing cue list means an unfiltered list stops being usable fast. Look at how `src/app/library/` does favourites/tags/summaries for instruments before inventing anything.

## 4. Spoken instructions — cached, never live TTS

**Cached audio only.** No runtime TTS calls, no API key in the browser, no per-run cost. Generate the clips offline, commit them, play them back.

- **Generate with `braidio`** (the ecosystem's voice/dialogue façade — `$PP/t/braidio`), not by calling ElevenLabs directly. `ELEVENLABS_API_KEY` is present in the dev environment. Per the ecosystem policy, use the local package and add any missing capability *there*.
- Pick **one** high-quality female voice on ElevenLabs and use it for every clip — a guidance voice that changes between cues is disorienting. Record the voice id in the generator script so the set is reproducible.
- A generator script (`scripts/gen_cue_voice.ts` or a Python sibling) takes the cue collection's instruction strings and emits one audio file per clip. It must be **idempotent and content-addressed**: hash the text, skip regeneration when unchanged. Otherwise every cue edit re-bills the whole set.
- Clips live in `public/` and ship with the bundle. Keep them small (Opus/AAC); the whole set should stay well under a megabyte.
- **Voice is a toggle; text is not.** Written instructions are always shown, and they are *also* dynamic — the spoken and written channels say the same thing at the same time. A player with sound off must lose nothing.

## 5. Dynamic guidance — the real feature

This is the part that makes it more than a script. The runner does not step through cues on a timer; it steps when it has **enough**.

```
SufficiencyEvaluator: (samples, cue, model?) => 
  | { verdict: 'need-more' }
  | { verdict: 'need-variation'; say: string }
  | { verdict: 'enough' }
  | { verdict: 'cannot'; why: string }
```

A strategy seam, so the first implementation can be simple (sample count + spread + coverage of the declared axes) and a smarter one can replace it without touching the runner. The `'cannot'` branch matters: if a player physically cannot produce what a cue asks for, the runner must say so and move on rather than looping forever.

Sequence in practice: *"Look to your left."* → enough → *"Now to your right."* → spread too small → *"A bit further, if you can."* → enough → *"Look up."*

**Design constraint:** every verdict must be reachable from data the sampler already produces. An evaluator needing something not captured is a spec bug, not a TODO.

## 6. Recording the session — reuse, do not rebuild

`src/app/recording/` already does all of this and is tested. Reuse:

- `SessionRecorder` (`session.ts`), `planRecording` (`plan.ts`), `buildManifest` (`manifest.ts`), `createRecordingSink` (`sink.ts`), `naming.ts`, `formats.ts`.
- The video must be the **clean camera stream** — no overlay, no text. `plan.ts` already distinguishes `pureVideo` from `overlayVideo`; use `pureVideo`.
- **Annotations**: `src/app/tagging/` and `src/taglog/` already exist for exactly this (#92). A cue becoming active is a tag; so is a sufficiency verdict. Emit them into the existing tag stream rather than inventing a parallel annotation format — the exporters (Audacity labels, WebVTT, CSV) then come free.
- Downloadable: the sink already supports the downloads location.

## 7. The projection view

**Start offline, but the library choice makes live nearly free.**

The decisive finding: **t-SNE is non-parametric and has no out-of-sample transform** — projecting a new live point means re-running it, or training a parametric t-SNE (a neural net). **UMAP does have one.** [`umap-js`](https://github.com/PAIR-code/umap-js) (PAIR-code, **MIT** — verify the LICENSE file, per policy) exposes `fit(data)` then `transform(newData)`.

So the architecture is one path, not two:

1. **Offline**: `fit()` the recorded take's still-points → a 2-D layout. Build and ship this first.
2. **Live cursor**: `transform()` the current live feature vector each frame → the cursor's position, in the *same* embedding. This is the same code, not a rewrite.

Do the `transform()` off the audio/tick path (it is human-frequency, like the trainer's ~30 Hz poll) and never `await` it in a tick.

**Point thinning.** The visualization literature calls this overdraw. Options in rough order of cost: alpha blending; hex/grid binning; **representative points that grow with the count they stand for** (what the maintainer described, and it is a real technique — cluster-based visual abstraction); Splatterplots (density-bounded abstraction with contours). Start with binned representatives; keep the cap a parameter.

## 8. Labelling in the projection → the category model

The payoff, and the step that is the same whichever projection produced the layout:

- View the projected points.
- **Select regions / points and label them.** Those labelled sets become the categories.
- This is an *alternative* to `cutAt(k)`, not a replacement — keep the automatic cut as the starting suggestion the player then edits. Selecting in 2-D is how a person fixes what the clustering got wrong.
- Categories must still be defined in the **full feature space**, not in 2-D: a label selects *points*, and the centroid is computed over their original vectors. A 2-D centroid would be meaningless to the classifier. This is the single easiest thing to get wrong here.

## 9. Gotchas for whoever picks this up

- **Reachability is a shipping rule.** New surface → `src/app/tools.ts` + a mount in `App.tsx` + a test in `test/tools_shell.test.tsx`. `test/app_shell.test.ts` will fail you if you forget the mount.
- **Licenses: permissive only**, and verify the LICENSE file rather than the npm field.
- **Gates**: `npm run typecheck` / `npm test` / `npm run build`. No ESLint. `src/enroll` is in `tsconfig.dag.json` — keep new dirs in the include list or they are not checked at all.
- **`npm run catalog`** after any node/param change; commit the generated files.
- **Headless testing is possible and expected**: `test/fixtures/video_head_pose/` is a real 405-frame recording with a written ground truth, including a camera-distance confound segment.
- **Deploy**: thoremin's own "Deploy" workflow only *dispatches*; the real run is in `tw_platform` and takes ~4 min. Verify the live `x-deploy-sha` against `git rev-parse HEAD` — and for a bundle change, that the served JS actually contains the new code.

Refs #160 (v1), #146 (live verification), #131 (invariance vocabulary), #92 (annotations), #88 (recording v2), #119 (feature catalog).

https://claude.ai/code/session_01FiZMdBN2v3u4FqaJFhK2Wi

**Comment by thorwhalen** [comment]
## Trainer v2 is implemented — five PRs, all merged

The redesign in this issue shipped as a chain of small, individually-reviewed PRs (each
went through a multi-agent adversarial review; none were stacked):

| PR | What |
|---|---|
| #164 | prep — the **feature-demand seam** (the vector nodes emit `{}` with the Lab closed; the trainer now claims groups so the stream is actually computed), the ms-clock fix, the face-model gate. The bug that meant v1 sampled nothing on a real webcam. |
| #170 | **Cues as a zodal collection** (`src/enroll/cue.ts`) + the **runner** with the `SufficiencyEvaluator` strategy seam + **noise units** (`noise.ts`) — every distance in multiples of a feature's own jitter, which is what makes a head-turn in degrees and a smile in 0..1 comparable. `ritual.ts` deleted; the maths (sampler/cluster/invariance/classify) kept. |
| #171 | **Guidance on the video** (the `trainerHud` overlay element, a per-device pref), the **routine picker** (filter + tags + save), the `GuidanceSink` seam. |
| #172 | **Cached spoken guidance** via braidio — one female voice (Sarah), content-addressed, `public/voice/`, a toggle. `npm run voice` regenerates; a coverage test fails if a cue is reworded without it. |
| #173 | **Record the take** — reuse `src/app/recording` (the clean `pureVideo` + `features.jsonl`) with **cue/verdict annotations** into the existing taglog stream, via a recording-controller seam on `useEngine`. |
| #174 | The **UMAP projection** + **draw-your-own-categories** — `fit()` offline, `transform()` the live cursor into the same embedding, select points, name groups. Categories are computed in **full feature space** (`session.modelFor`), never from 2-D. |

Everything is defined over `FeatureVector` in `src/enroll/` (no face-specific type — a hand
routine is a different list of group ids, not a different type); host glue is in
`src/app/enroll/`.

### The naming (from §1): **cue** and **routine**
A cue is one thing the trainer asks; a routine is an ordered list of cues. Both are
Zod-schema'd records in zodal collections.

### What is NOT done: the live webcam run
The pipeline is proven headlessly against the recorded `video_head_pose` fixture, but
**nobody has run the trainer with a real face** — the dwell/stillness defaults (220 ms,
"held" at 5σ) were set from that clip. A v2 verification checklist is posted on #146 §8
(the v1 one is obsolete). The thresholds are all parameters; the only reason they aren't
tuned is the missing live run.

Closing is the maintainer's call once that run is done.

**Comment by thorwhalen** [comment]
Housekeeping question for whoever owns this thread: **is #160 still doing any work?**

Its core shipped in PR #162, and #163 supersedes its design (cues/routines replacing the v1 ritual, per the vocabulary note now in CLAUDE.md). It came up in a backlog audit as the one open issue that looks finished but was never closed — flagging rather than closing it myself, since the trainer is your area and you may still be using it to track v1 leftovers.

If it is done, closing it leaves #163 as the single trainer thread.

**Comment by thorwhalen** [comment]
Answering the housekeeping question above ("is #160 still doing any work?"): yes, keep it open. #160 was retitled to cover only the "bind" verb — connecting a trained category to sound/dial output — which is a real, unshipped gap (verified at HEAD: `TrainedModel` still does not appear anywhere under `src/nodes/`, `src/music/`, `src/app/graph.ts`, `src/app/dials/` or `src/app/commands/`, and `createCategoryTracker` still has no caller outside `test/enroll.test.ts`). That is a distinct, still-open design decision (three readings laid out in #160's last comment — dispatcher vs dial vs graph node — plus the unresolved model-persistence prerequisite), not overlap with #163's now-shipped trainer v2. Closing #160 would lose that thread.
